Update SDR and NDR natcap/invest#555 (#37)
natcap/invest.users-guide: source/sdr.rst

where

 * :math:`R_i` is rainfall erosivity (units: :math:`MJ\cdot mm (ha\cdot hr\cdot yr)^{-1})`,

>  * :math:`K_i` is soil erodibility (units: :math:`ton\cdot ha\cdot hr (MJ\cdot ha\cdot mm)^{-1}`),

Reviewer: Are the units of erodibility correct _(hours_ versus _years_)? Why is _ha_ in both the denominator and the numerator when they would cancel out?

Reviewer 2: @jagoldstein, the units for erosivity are weird. As I commented previously: The correct units are listed in the Data Needs section: (MJ-mm)/(hour-ha-year). Also see this document, which walks through the gnarly units involved in the USLE: https://www.engr.colostate.edu/~pierre/ce_old/Projects/linkfiles/USLE%20Unit%20conversions.pdf.

Reviewer: Thanks, yeah the units get weird for both erodibility and erosivity, and I see different variations across different sources. I was specifically checking on _soil erodibility_ units here, which Emily did not change the definition of yet, but I just wanted to confirm that what's written for that is accurate.

Author: It's strange that the _ha_s don't cancel, but that paper @newtpatrol linked says:
> Although hectares  in  the numerator cancels hectares in the denominator,  both  are left  to show that  K  is  soil loss  per  unit area per unit of  E I .   
It also shows _hours_ for erodibility (see the table on page 3)

The value of :math:`m`, the length exponent of the LS factor, is based on the classical USLE, as discussed in (Oliveira et al., 2013):

.. math::

   \begin{align*}
   m &=  \left\{\begin{array}{lr}
      0.2, & \text{where } \theta \leq 1\% \\
      0.3, & \text{where } 1\% < \theta \leq 3.5\% \\
      0.4, & \text{where } 3.5\% < \theta \leq 5\% \\
      0.5, & \text{where } 5\% < \theta \leq 9\% \\
      \beta / (1 + \beta), & \text{where } \theta > 9\%
   \end{array}\right\} \\
   \\
   \beta &= \frac{\sin\theta / 0.0896}{3\sin\theta^{0.8} + 0.56}
   \end{align*}

> 

Reviewer: ```suggestion

```

.. _pour points:

Pour Points
+++++++++++

:math:`d_i` is the distance that water travels downslope from the pixel before reaching a stream. Because SDR is calculated in terms of :math:`d_i`, it is only defined for pixels that drain into a stream on the map.

In order to ensure that every pixel drains to a stream, and thus has defined :math:`d_i` and SDR values, the model adds pour points to the streams map if needed. A pour point is a point where water flows off the defined area of the flow direction raster, either off the edge or into a nodata area. If these points are not already marked as streams in the streams layer generated from flow accumulation (**flow_accum_streams.tif**) or the optional drainage layer input, the model will mark them as streams in **streams.tif**. These extra stream pixels may be a sign that something is wrong with your inputs:

- If the pour point is outside your watersheds of interest, you may ignore it.

> - If the pour point is actually a stream in the real world, try increasing your TFA and/or using the optional drainage layer so that the model recognizes this stream.

Reviewer: I'm confused here. I thought this was a list of how to deal with the model defining more pixels as streams than it should, so why would the user make a change "_so that the model recognizes this [more] stream_"? And increasing the TFA would lead the model to recognize fewer stream pixels anyway, not more.

Author: You're right it should say _decreasing_ TFA, not increasing, to make more streams. Hooray for reviews! 🎉 

My understanding is that, the user should provide data such that every pixel in their DEM drains into a stream.

When that doesn't happen, as of 3.9.1 the model will handle it by adding stream pixels on the pour points. But really, every pour point should _already_ be marked as a stream (either in the the stream layer that the model generates, and/or the optional streams input layer). 

Therefore, added stream pixels are a sign that you should mark more streams in that area, either by lowering the TFA or providing the optional streams input layer. Does that sound right?

Reviewer: Ok, that makes more sense to me. @newtpatrol do you agree?

> - If the pour point is not a stream in the real world, try using larger input data to be sure it completely covers the watershed including its streams.

Reviewer: What are "_larger input data_"? Do you mean "_input data with a greater extent_"?

Reviewer: ```suggestion
- If the pour point is not a stream in the real world, try inputting data with a greater spatial extent to be sure it completely covers the watershed including its streams.

```

Author: That does read better, thanks!

.. math:: R_i=dR_i\cdot\left(\left(\sum_{j\in\{pixels\ that\ drain\ to\ i\}}F_j \cdot p(i,j)\right) + E'_i\right)
    :label: ri

where :math:`F_j` is the amount of sediment-export-that-does-not-reach-the stream "flux", defined as:

.. math:: F_i=(1-dR_i)\cdot\left(\left(\sum_{j\in\{pixels\ that\ drain\ to\ i\}} F_j \cdot p(i,j)\right) + E'_i\right)
    :label: fi

Optional Drainage Layer
^^^^^^^^^^^^^^^^^^^^^^^

In some situations, the index of connectivity defined by topography does not represent actual flow paths, which may be influenced by artificial connectivity instead. For example, sediments in urban areas or near roads are likely to be conveyed to the stream with little retention. The (optional) drainage raster identifies the pixels that are artificially connected to the stream, irrespective of their geographic position (e.g. their distance to the stream network). Pixels from the drainage layer are treated similarly to pixels of the stream network; in other words, the downstream flow path will stop at pixels of the drainage layer (and the corresponding sediment load will be added to the total sediment export).

Defined Area of Outputs
^^^^^^^^^^^^^^^^^^^^^^^

> SDR and several other model outputs are defined in terms of distance to stream (:math:`d_i`). Therefore, these outputs are only defined for pixels that drain to a stream on the map (and so are within the streams' watershed). Pixels that do not drain to any stream will have nodata in these outputs. Increasing the threshold flow accumulation value may shrink the defined area of these outputs, as fewer stream pixels exist and the watershed is smaller. Correspondingly, decreasing the threshold flow accumulation value may expand the defined area of these outputs.

Reviewer 2: This is very confusing to me. I have never seen increasing the TFA shrink the output area, it just creates fewer tributaries. Increasing the TFA has never expanded the area, it just creates more tributaries. If it did "expand", it couldn't go past the boundary of the input data. I'm guessing that this verbiage is related to removing the pour points? If so, instead of the shrink/expand text that's here, I'd suggest further explaining why they might not drain to a stream (going off the map, or pits/errors in the DEM), and describing what the output will look like for pixels that do not drain to the stream (you'll see areas of NoData that can't be explained by missing data in your input rasters...) While I'm thinking about this, was it crazy to suggest that these pixels receive a distinctly bogus value like -9999, so they'd be really obvious to people, and they wouldn't have to pour through their inputs to see what's causing NoData in the output?

Author: Yes, this is new behavior related to removing the pour points. I tried to revise the text as you suggest and added an image comparing the stream and SDR outputs with different TFA values. I'm struggling to explain this effectively.

Having a separate "nodata" value like -9999 could definitely be useful, and I made a ticket for it: #621 However it would require more not-insignificant code changes. 3.9.1 is just a bugfix release and in the interest of getting it out, I think we should come back to that idea as a possible feature in 3.10 or so.

Reviewer 2: Thanks for working on this @emlys!  I'm about to take off for the day, but will take a look at the graphics, and would be happy to help word the explanation.

And thanks for considering the value-other-than-NoData for disconnected pixels. No hurry, just an idea.

Author: Thanks @newtpatrol!

> 

Reviewer: ```suggestion

```

In some situations, the index of connectivity defined by topography does not represent actual flow paths, which may be influenced by artificial connectivity instead. For example, sediments in urban areas or near roads are likely to be conveyed to the stream with little retention. The (optional) drainage raster identifies the pixels that are artificially connected to the stream, irrespective of their geographic position (e.g. their distance to the stream network). Pixels from the drainage layer are treated similarly to pixels of the stream network; in other words, the downstream flow path will stop at pixels of the drainage layer (and the corresponding sediment load will be added to the total sediment export).

> 

Reviewer: ```suggestion

```

Defined Area of Outputs
^^^^^^^^^^^^^^^^^^^^^^^

SDR and several other model outputs are defined in terms of distance to stream (:math:`d_i`). Therefore, these outputs are only defined for pixels that drain to a stream on the map (and so are within the streams' watershed). Pixels that do not drain to any stream will have nodata in these outputs. Increasing the threshold flow accumulation value may shrink the defined area of these outputs, as fewer stream pixels exist and the watershed is smaller. Correspondingly, decreasing the threshold flow accumulation value may expand the defined area of these outputs.

The affected output files are: **d_dn.tif**, **ic.tif**, **e_prime.tif**, **sdr_factor.tif**, **sdr_bare_soil.tif**, **d_dn_bare_soil.tif**, **ic_bare_soil.tif**, **sed_retention.tif**. **sed_retention_index.tif**, **sediment_deposition.tif**, and **sed_export.tif**

The model's stream map is the union of the calculated stream layer and the input drainage layer (if provided).

> The model calculates a stream layer (**stream.tif**) by thresholding the flow accumulation raster (**flow_accumulation.tif**) by the threshold flow accumulation value:

Reviewer: ```suggestion
The model calculates a stream layer (**stream.tif**) by thresholding the flow accumulation raster (**flow_accumulation.tif**) by the threshold flow accumulation value (TFA):

```

  .. math::
     :label: sdr_stream

     stream_{TFA,i} = \left\{\begin{array}{lr}
          1, & \text{if } flow\_accum_{i} \geq TFA \\
          0,     & \text{otherwise} \\
          \end{array}\right\}

If the optional drainage input is provided, the model includes it (**stream_and_drainage.tif**):

  .. math:: stream_{drainage,i} = stream_{TFA,i} \text{  OR  } stream_{input,i}
     :label: stream_and_drainage

The final stream layer (:math:`stream_{TFA}`, or :math:`stream_{drainage}` if the optional drainage input is provided) is used to determine :math:`d_i` for the SDR calculations.

> 

Reviewer: ```suggestion

```

> 

Reviewer: ```suggestion

```
